File: app/src/main/java/com/teambrella/android/ui/teammate/TeammateObjectFragment.java
package com.teambrella.android.ui.teammate;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.load.resource.bitmap.CenterCrop;
import com.bumptech.glide.load.resource.bitmap.RoundedCorners;
import com.bumptech.glide.request.RequestOptions;
import com.google.gson.JsonElement;
import com.google.gson.JsonObject;
import com.teambrella.android.R;
import com.teambrella.android.api.TeambrellaModel;
import com.teambrella.android.api.model.json.JsonWrapper;
import com.teambrella.android.image.glide.GlideApp;
import com.teambrella.android.ui.base.ADataFragment;
import com.teambrella.android.ui.claim.ClaimActivity;
import com.teambrella.android.ui.image.ImageViewerActivity;
import com.teambrella.android.ui.team.claims.ClaimsActivity;
import com.teambrella.android.util.AmountCurrencyUtil;

import java.util.ArrayList;

import io.reactivex.Notification;
import io.reactivex.Observable;

/**
 * Teammate Object Fragment
 */
public class TeammateObjectFragment extends ADataFragment<ITeammateActivity> {

    private ImageView mObjectPicture;
    private TextView mObjectModel;
    private TextView mLimit;
    private TextView mNet;
    private TextView mRisk;
    private TextView mSeeClaims;
    private TextView mCoverageType;
    private TextView mObjectTitle;

    private int mTeammateId;
    private int mTeamId;
    private int mClaimId;
    private int mClaimCount;
    private String mModel;
    private String mCurrency;
    private int mGender;


    public static TeammateObjectFragment getInstance(String dataTag) {
        TeammateObjectFragment fragment = new TeammateObjectFragment();
        Bundle args = new Bundle();
        args.putString(EXTRA_DATA_FRAGMENT_TAG, dataTag);
        fragment.setArguments(args);
        return fragment;
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_teammate_object, container, false);
        mObjectModel = view.findViewById(R.id.model);
        mObjectPicture = view.findViewById(R.id.object_picture);
        mLimit = view.findViewById(R.id.limit);
        mNet = view.findViewById(R.id.net);
        mRisk = view.findViewById(R.id.risk);
        mSeeClaims = view.findViewById(R.id.see_claims);
        mCoverageType = view.findViewById(R.id.coverage_type);
        mObjectTitle = view.findViewById(R.id.object_title);
        return view;
    }


    @Override
    protected void onDataUpdated(Notification<JsonObject> notification) {
        if (notification.isOnNext()) {

            Observable<JsonWrapper> responseObservable = Observable.just(notification.getValue())
                    .map(JsonWrapper::new);

            Observable<JsonWrapper> dataObservable =
                    responseObservable.map(response -> response.getObject(TeambrellaModel.ATTR_DATA))
                            .doOnNext(data -> mTeammateId = data.getInt(TeambrellaModel.ATTR_DATA_ID, mTeammateId));

            Observable<JsonWrapper> basicObservable =
                    dataObservable.map(jsonWrapper -> jsonWrapper.getObject(TeambrellaModel.ATTR_DATA_ONE_BASIC));

            basicObservable.doOnNext(jsonWrapper -> mGender = jsonWrapper.getInt(TeambrellaModel.ATTR_DATA_GENDER, mGender))
                    .onErrorReturnItem(new JsonWrapper(null)).blockingFirst();

            Observable<JsonWrapper> objectObservable =
                    dataObservable.map(item -> item.getObject(TeambrellaModel.ATTR_DATA_ONE_OBJECT));

            dataObservable.map(jsonWrapper -> jsonWrapper.getObject(TeambrellaModel.ATTR_DATA_ONE_TEAM))
                    .doOnNext(jsonWrapper -> mCoverageType.setText(TeambrellaModel.getInsuranceTypeName(jsonWrapper.getInt(TeambrellaModel.ATTR_DATA_COVERAGE_TYPE))))
                    .doOnNext(jsonWrapper -> mCurrency = jsonWrapper.getString(TeambrellaModel.ATTR_DATA_CURRENCY, mCurrency))
                    .doOnNext(jsonWrapper -> mObjectTitle.setText(mDataHost.isItMe() ? TeambrellaModel.getMyObjectNamer(jsonWrapper.getInt(TeambrellaModel.ATTR_DATA_COVERAGE_TYPE))
                            : TeambrellaModel.getObjectNameWithOwner(jsonWrapper.getInt(TeambrellaModel.ATTR_DATA_COVERAGE_TYPE), mGender)))
                    .onErrorReturnItem(new JsonWrapper(null))
                    .blockingFirst();


            objectObservable.doOnNext(objectData -> mObjectModel.setText(objectData.getString(TeambrellaModel.ATTR_DATA_MODEL)))
                    .doOnNext(objectData -> AmountCurrencyUtil.setAmount(mLimit, Math.round(objectData.getFloat(TeambrellaModel.ATTR_DATA_CLAIM_LIMIT)), mCurrency))
                    .doOnNext(objectData -> mClaimId = objectData.getInt(TeambrellaModel.ATTR_DATA_ONE_CLAIM_ID, mClaimId))
                    .doOnNext(objectData -> mModel = objectData.getString(TeambrellaModel.ATTR_DATA_MODEL, mModel))
                    .doOnNext(objectData -> mClaimCount = objectData.getInt(TeambrellaModel.ATTR_DATA_CLAIM_COUNT, mClaimCount))
                    .onErrorReturnItem(new JsonWrapper(null)).blockingFirst();

            ArrayList<String> photos = objectObservable.flatMap(objectData -> Observable.fromIterable(objectData.getJsonArray(TeambrellaModel.ATTR_DATA_SMALL_PHOTOS)))
                    .map(JsonElement::getAsString)
                    .toList(ArrayList::new)
                    .onErrorReturn(throwable -> new ArrayList<>()).blockingGet();

            if (photos != null && photos.size() > 0) {
                Context context = getContext();
                GlideApp.with(this).load(getImageLoader().getImageUrl(photos.get(0)))
                        .apply(new RequestOptions().transforms(new CenterCrop(), new RoundedCorners(getResources().getDimensionPixelOffset(R.dimen.rounded_corners_4dp))
                        ).placeholder(R.drawable.picture_background_round_4dp))
                        .into(mObjectPicture);
                mObjectPicture.setOnClickListener(v -> v.getContext().startActivity(ImageViewerActivity.getLaunchIntent(context, photos, 0)));
            }

            basicObservable.doOnNext(basic -> AmountCurrencyUtil.setAmount(mNet, Math.round(basic.getFloat(TeambrellaModel.ATTR_DATA_TOTALLY_PAID_AMOUNT)), mCurrency))
                    .doOnNext(basic -> mRisk.setText(getString(R.string.risk_format_string, basic.getFloat(TeambrellaModel.ATTR_DATA_RISK) + 0.05f)))
                    .doOnNext(basic -> mTeamId = basic.getInt(TeambrellaModel.ATTR_DATA_TEAM_ID, mTeamId))
                    .onErrorReturnItem(new JsonWrapper(null)).blockingFirst();

            mSeeClaims.setEnabled(mClaimCount > 0);
            mSeeClaims.setText(getContext().getString(R.string.see_claims_format_string, mClaimCount));

            mSeeClaims.setOnClickListener(v -> {
                if (mClaimId > 0) {
                    ClaimActivity.start(getContext(), mClaimId, mModel, mTeamId);
                } else {
                    ClaimsActivity.start(getContext(), mTeamId, mTeammateId, mCurrency);
                }
            });
        }

    }
}

File: app/src/main/java/com/teambrella/android/ui/teammate/TeammateObjectFragment.kt
package com.teambrella.android.ui.teammate

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageView
import android.widget.TextView
import com.google.gson.JsonObject
import com.teambrella.android.R
import com.teambrella.android.api.*
import com.teambrella.android.ui.base.AKDataFragment
import com.teambrella.android.ui.claim.ClaimActivity
import com.teambrella.android.ui.image.ImageViewerActivity
import com.teambrella.android.ui.team.claims.ClaimsActivity
import com.teambrella.android.ui.util.setAVGDifference
import com.teambrella.android.ui.util.setImage
import com.teambrella.android.util.AmountCurrencyUtil
import io.reactivex.Notification

class TeammateObjectFragment : AKDataFragment<ITeammateActivity>() {

    private val objectModel: TextView? by ViewHolder(R.id.model)
    private val objectPicture: ImageView? by ViewHolder(R.id.object_picture)
    private val limit: TextView? by ViewHolder(R.id.limit)
    private val net: TextView? by ViewHolder(R.id.net)
    private val risk: TextView? by ViewHolder(R.id.risk)
    private val seeClaims: TextView? by ViewHolder(R.id.see_claims)
    private val coverageType: TextView? by ViewHolder(R.id.coverage_type)
    private val objectTitle: TextView? by ViewHolder(R.id.object_title)
    private val avgDifference: TextView? by ViewHolder(R.id.avg_difference)


    private var teammateId: Int = 0
    private var teamId: Int = 0
    private var claimId: Int = 0
    private var claimCount: Int = 0
    private var model: String? = null
    private var currency: String? = null
    private var gender: Int = 0

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_teammate_object, container, false)
    }

    override fun onDataUpdated(notification: Notification<JsonObject>?) {
        val data = notification?.takeIf { it.isOnNext }?.value?.data
        data?.let { _data ->
            teammateId = _data.intId ?: teammateId
            gender = _data.basic?.gender ?: gender
        }

        data?.team?.let { _team ->
            coverageType?.setText(TeambrellaModel.getInsuranceTypeName(_team.coverageType ?: 0))
            currency = _team.currency ?: currency
            objectTitle?.setText(
                    if (mDataHost.isItMe)
                        TeambrellaModel.getMyObjectNamer(_team.coverageType ?: 0)
                    else
                        TeambrellaModel.getObjectNameWithOwner(_team.coverageType ?: 0, gender))
        }

        data.objectPart?.let { _object ->
            model = _object.model ?: model
            objectModel?.text = model
            AmountCurrencyUtil.setAmount(limit, Math.round(_object.claimLimit ?: 0f), currency)
            claimId = _object.oneClaimId ?: claimId
            claimCount = _object.claimCount ?: claimCount

            val photos = ArrayList<String>().apply {
                _object.smallPhotos?.let {
                    addAll(it.map { it.asString })
                }
            }
            if (photos.size > 0) {
                objectPicture?.setImage(imageLoader.getImageUrl(photos[0]), R.dimen.rounded_corners_4dp, R.drawable.picture_background_round_4dp)
                objectPicture?.setOnClickListener({ startActivity(ImageViewerActivity.getLaunchIntent(context, photos, 0)) })
            }

        }

        data.basic?.let { _basic ->
            AmountCurrencyUtil.setAmount(net, Math.round(_basic.totallyPaidAmount ?: 0f), currency)
            risk?.text = getString(R.string.float_format_string_2, _basic.risk ?: 0f)
            avgDifference?.setAVGDifference(_basic.risk ?: 0.2f, _basic.averageRisk ?: 1f)
            teamId = _basic.teamId ?: teamId
        }


        seeClaims?.isEnabled = claimCount > 0
        seeClaims?.text = getString(R.string.see_claims_format_string, claimCount)
        seeClaims?.setOnClickListener({
            if (claimId > 0) {
                ClaimActivity.start(context!!, claimId, model, teamId)
            } else {
                ClaimsActivity.start(context!!, teamId, teammateId, currency)
            }
        })
    }
}